Display labels instead of icons in top toolbar.

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #10524 , along with #24234 . This PR points to #24234 so that only the changeset related to displaying button labels in the top toolbar is visible.

All the feedback and comments from #15830 have been addressed. One point that I'd appreciate further feedback on is the positioning of nested dropdowns, such as:

<img width="630" alt="Screen Shot 2020-07-31 at 3 45 49 pm" src="https://user-images.githubusercontent.com/8096000/89006315-d0f4d100-d349-11ea-8e3f-e030a874dfbd.png">


## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested in browser at different breakpoints. Keyboard navigation checked.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

tellthemachines committed:
Add option to show icon labels.

tellthemachines committed:
Address review feedback and consolidate

tellthemachines committed:
Move option to general section

tellthemachines committed:
Display labels instead of icons in top toolbar.

mapk commented:
Thanks for this PR! I spent some time working with the [previous design explorations](https://github.com/WordPress/gutenberg/pull/15830) and wanted to present my further iterations here. While this isn't direct feedback on the PR (yet), I thought it best to share on this PR for a discussion as to what path works best.

----
### Option 1 - Add labels under icons
![txt-under-icons](https://user-images.githubusercontent.com/617986/89237390-e6673680-d5a7-11ea-8144-19a5cfbee080.png)

- Reduce the active state of the Settings icon from 36px to 32px. This allows more room for text and aligns it to the same size as the Inserter.
----
### Option 2 - Expand height of topbar
![76px](https://user-images.githubusercontent.com/617986/89237553-55dd2600-d5a8-11ea-9444-5023230345bc.png)

- Height has been expanded from 60px to 76px to allow more room for text.
- The "W", at this size, feels too big to be full bleed. I dropped it down a bit and pulled it away from edges.
- The right side buttons feel a little out of alignment and would need further iterations.
----
### Option 3 - Smaller icons
![smaller-icons](https://user-images.githubusercontent.com/617986/89237674-a9e80a80-d5a8-11ea-960e-4a3d258ce7f1.png)

**Mobile**

![Screen Shot 2020-08-03 at 4 47 00 PM](https://user-images.githubusercontent.com/617986/89237800-fb909500-d5a8-11ea-90c3-bbe721562df1.png)

- I resized the icons from 36px to 28px. This doesn't align with our standard icon size, but does allow for more breathing room when icon labels are visible.
- With the smaller icons, the positioning of the Publish button doesn't feel so bad.
- Again, the Settings icon active size matches the Inserter size.
----
### Option 4 - Text only
![Screen Shot 2020-08-03 at 4 48 42 PM](https://user-images.githubusercontent.com/617986/89237881-398db900-d5a9-11ea-943b-899e8da57126.png)

**Mobile**

![mobile](https://user-images.githubusercontent.com/617986/89238182-11eb2080-d5aa-11ea-8c48-81d0d51f2277.gif)


- Eliminate all icons (as this PR does already).
- I've explored several iterations with the Mover text.
- As shown in the Mobile view, it would be better to display the block's name in the toolbar rather than just "block".

----

### Feedback
I wanted to get these early iterations up for discussion. Is there an option that feels better than the others? What else would you like to see iterated? If you're interested in more, check out my [Figma file](https://www.figma.com/file/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=0%3A1).

noisysocks commented on packages/components/src/dropdown-menu/index.js:
```diff
@@ -129,7 +129,7 @@ function DropdownMenu( {
 						aria-expanded={ isOpen }
 						label={ label }
-						showTooltip
+						showTooltip={ toggleProps?.showTooltip || true }
```
This will always set `showTooltip` regardless of what `toggleProps?.showTooltip` is.

Assuming that `true` is supposed to be a default value that can be overridden by `toggleProps?.showTooltip`, we could use `??` instead of `||`:

```suggestion
						showTooltip={ toggleProps?.showTooltip ?? true }
```

tellthemachines replied:
Oooh, well spotted! Fixing.

noisysocks reviewed (commented):
Nice job! Seems to work pretty well. Waiting for design to be finalised before I test thoroughly.

tellthemachines committed:
Remove unnecessary dispatch

mapk commented:
Here's my feedback on the current state of this PR:

### Text only
I love the text only mode because it's so clean. But once the Top Toolbar option is introduced, it feels a little wonky. This PR chooses to keep the block toolbar on a second row, which could work, but would need updating to text only as well. (I think that's reserved for another PR)

-----

### Top Toolbar mode
The bottom border is missing in the second level toolbar here:

![Screen Shot 2020-08-04 at 4 04 00 PM](https://user-images.githubusercontent.com/617986/89353880-3c051700-d66c-11ea-9c1d-e1142c6adfce.png)

-----

### Site icon
I don't think we need to add text here. It just gets very out of place and messy.

![Screen Shot 2020-08-04 at 3 58 38 PM](https://user-images.githubusercontent.com/617986/89354145-ef6e0b80-d66c-11ea-93d6-c4c649d1d14b.png)

-----

### Text edits
- "Add block" should just be "Add"
- "Block navigation" should just be "List view"

-----

### Settings button
This looks more like a button now than a toggle now. Should keep this pattern?

-----

### Button sizes
When both menu dropdowns are triggered, the "Add block" and "Publish" btns sizes are more visibly unequal and look like a mistake. I think they should all match sizes with this option.

![Screen Shot 2020-08-04 at 4 14 29 PM](https://user-images.githubusercontent.com/617986/89354483-d44fcb80-d66d-11ea-9f09-acfee7cc8f89.png)

-----

### Dropdown menus should not overlap
These dropdowns should appear next to their parent level menu.

**Example 1**

![menu-drops](https://user-images.githubusercontent.com/617986/89354769-943d1880-d66e-11ea-938c-fc158fe14711.png)

**Example 2**

![drop2](https://user-images.githubusercontent.com/617986/89354893-e0885880-d66e-11ea-859e-b4988c2701a6.png)

-----

### Settings item when in dropdown
I think we should drop the "active" state of the Settings item when it's in the dropdown. More like this:

![Screen Shot 2020-08-04 at 4 26 29 PM](https://user-images.githubusercontent.com/617986/89355055-4a086700-d66f-11ea-9a06-0f1dff6a727c.png)

-----

### Dropdowns
I think we have a standard min-width on dropdowns like 258px. Let's apply that to these as well.

-----
After these changes, I'm happy to provide another review. 😄

tellthemachines committed:
Fix dropdown tooltip display

tellthemachines committed:
Adjust button and toolbar styling.

tellthemachines commented:
I've left this in labels-only mode for now, we can always change it later if we decide it doesn't work. Or we could have three options: icons only, icons and labels, and labels only 😄 

I've addressed most of @mapk 's feedback; all but the dropdown positioning as I'm still tussling with the Popover component on that front. Hopefully will have it fixed soon!

The e2e tests are failing because of the changes I made to the label text on some of the buttons. I'd appreciate any feedback on those changes, so that I can fix the tests once they're sorted.

mapk commented:
Thanks for the quick turnaround, @tellthemachines! As you're working through this, I hope providing a bit more feedback is okay. If you'd like me to hold off until a certain point, just let me know.

### Top Toolbar mode
I noticed that when selecting Top Toolbar mode, when I don't have a block selected the topbar bottom border doubles up. 

![border](https://user-images.githubusercontent.com/617986/89445653-4d9bfc80-d708-11ea-89b8-09cf099106ac.gif)

----

### Text buttons
When switching to a text only mode, I really think all the button interactions should work the same. It always felt odd to me to have icon buttons (ie. hover and pressed states) display differently than the text buttons, and now that these are all text buttons, it's even more obvious. 

**An open question:** Should the text only buttons reflect the same interactions that the Preview button has. These are no longer icon buttons, and should probably now reflect the interactions of our standard button states.

![btns](https://user-images.githubusercontent.com/617986/89445867-90f66b00-d708-11ea-8b8f-b3b7c8c22647.gif)

----

### Settings and Options
I just noticed that when both Settings and Options are in a dropdown, and activated, they both show blue. This doesn't feel right to me even though technically it is. Maybe fixing the whole button interaction thing will help this?

![Screen Shot 2020-08-05 at 10 46 11 AM](https://user-images.githubusercontent.com/617986/89447160-660d1680-d70a-11ea-887d-1aff69f4c55b.png)

----

### Icon dropdown
In text only mode, when the screen size is reduced, the buttons all move to a dropdown that is represented only with an icon. Does this make sense in the text only mode? Should we have a label for this icon?

![Screen Shot 2020-08-05 at 10 57 31 AM](https://user-images.githubusercontent.com/617986/89447402-c1d79f80-d70a-11ea-90bf-4f2bd1447952.png)


----
There are so many small nuances that come to light as this gets developed, so I really appreciate your time with it all.

mapk commented:
If we converted all the topbar text-only buttons to the default button instance as outlined in our Figma libraries, they would interact similarly to the Preview button. Check out this prototype below: (please ignore the black background)

![btns](https://user-images.githubusercontent.com/617986/89454639-da00ec00-d715-11ea-92d4-a34e427ee6b9.gif)

tellthemachines committed:
Reposition outline dropdown

tellthemachines committed:
Try button tertiary styles

tellthemachines commented:
> As you're working through this, I hope providing a bit more feedback is okay.

Absolutely, all feedback helps!

> I noticed that when selecting Top Toolbar mode, when I don't have a block selected the topbar bottom border doubles up.

Hmm, I can't reproduce this. Have you checked out my latest changes? 

> In text only mode, when the screen size is reduced, the buttons all move to a dropdown that is represented only with an icon. Does this make sense in the text only mode? Should we have a label for this icon?

Very good question 😅 
I'm happy to try out a text label here, just not sure what to call it (we don't currently have an aria-label on this button, so I guess that'll need addressing either way). "More settings" seems a bit weird/redundant? 

I've switched all the text buttons to behave like the Preview button in my latest commit, let me know if it works!

Regarding the dropdown positioning issues, it's trickier than I thought initially. The Popover component that we use in dropdowns has very limited positioning capabilities. We have #21275 to address that, and from the looks of it it's going to require rewriting the whole component. I repositioned the Outline dropdown the best I could, but haven't been able to improve any of the others. Can we leave it as is for now, and revisit once Popover is rebuilt?

mapk commented:
>I've switched all the text buttons to behave like the Preview button in my latest commit, let me know if it works!

After sitting on this one for a few days, I'm not fond of how our default button state looks for each of these topbar items. I iterated further with something more reminiscent of our Icon Button styles. It's not perfect, but the other direction felt overwhelming.

- The gray pressed state is replaced with a simple 1.5 pixel blue border.
- I aligned with the [recent PR](https://github.com/WordPress/gutenberg/pull/24420) to make the text links in the topbar blue instead of black.
- Hovered states are identified with a border as are the active states.

![icon-like-buttons](https://user-images.githubusercontent.com/617986/89834869-782af280-db18-11ea-93fa-c41b073641a3.gif)

### Next steps
- Iterate when the Top Toolbar mode is "on". How do the toolbar buttons look when there are gray dividers?
- Explore other interaction designs.

tellthemachines committed:
Update button styles

tellthemachines committed:
Add translator comments to top toolbar buttons.

tellthemachines committed:
Fix failing tests.

tellthemachines committed:
Update e2e test selector.

mapk commented:
I got some feedback yesterday that moving to "text only" might be too much. The text only version loses hierarchy and grouping which can cause usability issues. 

I'm exploring a couple icon + text solutions to see how else we might be able to retain some sense of hierarchy and relationship once we introduce text into the topbar.

cc @kellychoffman

ZebulanStanphill commented:
@mapk Could you elaborate on how is it losing hierarchy/grouping? I suspect the issue is not with text vs. icons, but with the pre-existing lack of clear division between unrelated controls in the top toolbar. Why aren't there visible dividers like how controls are grouped in the block toolbar? I've said it before, but the top toolbar doesn't seem to have a clear sense of where controls go. (And it's even worse with "Top Toolbar" enabled; the block toolbar should be treated as a separate toolbar below the top one.)

tellthemachines commented:
> @mapk Could you elaborate on how is it losing hierarchy/grouping?

I'm curious about this too. I'm not aware of any existing hierarchy or grouping (apart from left vs right hand side of the top toolbar), so perhaps that's something that needs to be made clearer in the overall design.

mapk commented:
@tellthemachines and @ZebulanStanphill You both are correct. While the icons lacked organization, they didn't take up a large amount of space, and the shapes helped differentiate them. The text-only option brings stronger focus to the lack of both their order and their level of hierarchy in relation to each other. Everything is now a similar size and importance. For example, the "redo" and "undo" items feel too prominent.

talldan commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -96,18 +140,25 @@ function HeaderToolbar() {
+						<Button
+							className="button-toggle"
+							aria-expanded={ isOpen }
+							icon={ chevronDown }
```
I know there was some discussion on the PR about the design of this, but should this also be a text button?

It seems inconsistent to have an icon button in this mode.

It also looks like it needs a label.

ZebulanStanphill replied:
This raises the question of what label to give this button. There are two of them, and just calling them both "More" seems a bit confusing, since there's no clear answer to "More what?" ...which I guess just loops back to the same issue we keep running into: the top toolbar is not organized... or rather, what organization it does have is very vague and arbitrary.

talldan replied:
Hmm, yep, Settings and Options are also often interchangeable, but they are in the same menu at small breakpoints with not much to distinguish them.

I do think there's a bit of a separation in the way options are grouped. The items on the left are more about interacting with content, while those on the right are more related to the editor features or publishing. Not sure I'm smart enough to dilute that into button labels though. The best I can come up with is:
- 'More content tools'
- 'More editor options'

tellthemachines replied:
Yeah, we seem to be going around in circles on this. "More tools and options" has already become "Options", and these dropdowns include that and a heap of other stuff that isn't easy to label as a group 😬

tellthemachines replied:
> The best I can come up with is:
> 'More content tools'
> 'More editor options'

That's much better than my 0 ideas so far 😅 
We'd need them to be shorter though. And clicking on an "options" button only to find another "options" button inside would be a bit 😭. 
Putting the sidebar toggle inside the options menu doesn't seem right either. I wonder if we could leave it out and instead pop the "switch to draft" in the "options"?

ZebulanStanphill replied:
Perhaps instead of having essentially a more menu inside of another more menu on the right, just have one more menu that stuff moves into as necessary (much like how the overflow menu in the header of the Twenty Nineteen theme works). The toolbar should also be reorganized so the most important controls are listed first, so they remain visible at more viewport widths, e.g. "Publish" should come before "Save draft" and "Preview".

talldan commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -68,11 +75,48 @@ function HeaderToolbar() {
+			/>
+			<ToolbarItem
+				as={ BlockNavigationDropdown }
+				isDisabled={ isTextModeEnabled }
```
The disabled items in the overflow menu don't seem to be consistent with other dropdown menus:
<img width="281" alt="Screenshot 2020-08-13 at 4 01 11 pm" src="https://user-images.githubusercontent.com/677833/90109827-ed97fc80-dd7e-11ea-9043-cabe6d7e48fc.png">

tellthemachines replied:
Yeah, they are meant to be consistent with their out-of-overflow state 😅 but perhaps @mapk can comment further on that.

talldan commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -68,11 +75,48 @@ function HeaderToolbar() {
+	const overflowItems = (
+		<>
+			{ isLargeViewport && (
+				<ToolbarItem
+					as={ ToolSelector }
+					showTooltip={ ! showIconLabels }
+					isTertiary={ showIconLabels }
+				/>
```
More of a question for design, I wonder if it could be worth bringing the tools inline in the overflow menu instead of having them in an additional popover.

So instead of having something like:
<img width="334" alt="Screenshot 2020-08-13 at 4 11 28 pm" src="https://user-images.githubusercontent.com/677833/90110464-da396100-dd7f-11ea-939e-f6b86a141f3a.png">


It could be more like the Visual Editor / Code Editor options in the more menu in a separate section of the overflow menu:
<img width="264" alt="Screenshot 2020-08-13 at 4 11 45 pm" src="https://user-images.githubusercontent.com/677833/90110505-ec1b0400-dd7f-11ea-8136-9adfde7b339b.png">

tellthemachines replied:
This relates to the earlier [conversation about hierarchy](https://github.com/WordPress/gutenberg/pull/24304#issuecomment-672964514). 
My only thought from a code standpoint is this will require quite a bit of markup/logic duplication, as we would have to use different components for each breakpoint.

mapk replied:
Separating these under section headings and using dividers could really help. I'll explore some ideas.

talldan commented on packages/edit-post/src/components/header/index.js:
```diff
@@ -61,8 +81,23 @@ function Header( { setEntitiesSavedStatesCallback } ) {
+						contentClassName="edit-post-header__dropdown"
+						position="bottom right"
+						renderToggle={ ( { isOpen, onToggle } ) => (
+							<Button
+								className="button-toggle"
+								aria-expanded={ isOpen }
+								icon={ chevronDown }
+								isSecondary
+								onClick={ onToggle }
+							/>
```
Similar again here, this button seems to need a label.

The button styles in the dropdown also look like they need some tweaking:
<img width="96" alt="Screenshot 2020-08-13 at 4 08 04 pm" src="https://user-images.githubusercontent.com/677833/90111086-b7f41300-dd80-11ea-8227-28be38e77e5b.png">

talldan reviewed (commented):
Great work on this, I can see it's not at all straightforward to wrangle all that text and make things work on various screen sizes.

Looks pretty close to ready, mostly just the dropdown menus need a bit more work so that they're consistent.

tellthemachines commented on packages/block-editor/src/components/block-navigation/dropdown.js:
```diff
@@ -47,6 +47,7 @@ function BlockNavigationDropdownToggle( {
 	);
 
 	return (
+		/* translators: button label text should, if possible, be under 16 characters. */
```
Ooooh right.

tellthemachines committed:
Fix translator comments.

tellthemachines commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -96,18 +140,25 @@ function HeaderToolbar() {
-			/>
+			{ ( isWideViewport || ! showIconLabels ) && overflowItems }
+			{ ! isWideViewport && ! isSmallViewport && showIconLabels && (
+				<Dropdown
```
I don't think I can use `DropdownMenu` here because its only permitted children are `MenuItem`, `MenuItemsChoice`, or `MenuGroup`, which don't seem right semantically. Unless I'm missing something?

tellthemachines committed:
Change header toolbar menu labels.

tellthemachines committed:
Change options labels

tellthemachines committed:
Update snapshot.

tellthemachines commented:
I've updated this PR to add text labels to the dropdowns that appear on smaller breakpoints. I had to make a couple of changes to existing labels in order to avoid duplication, so the left hand dropdown is called "Tools" and I changed the existing "Tools" dropdown to be called "Modes" as I think that describes what it does a bit better. The right hand dropdown is now called "Options" and I updated the "More tools & options" button to be called "More options".

This works well on very small breakpoints, even when plugins have registered buttons on the right hand side slot:

<img width="320" alt="Screen Shot 2020-08-25 at 12 02 30 pm" src="https://user-images.githubusercontent.com/8096000/91114319-f962b780-e6ca-11ea-872b-90b6e4b2ee20.png">

This is something that the current layout with the icon buttons doesn't solve very well, so perhaps we can use this as an exercise to consider further improvements to small breakpoint layouts:

<img width="322" alt="Screen Shot 2020-08-25 at 11 15 22 am" src="https://user-images.githubusercontent.com/8096000/91114408-2fa03700-e6cb-11ea-8332-4be6e8c82948.png">

tellthemachines committed:
Update e2e selector

mapk commented:
Taking another look! Here's a visual of the current state of this PR:

![current-state](https://user-images.githubusercontent.com/617986/91235862-4e162a80-e6eb-11ea-8eec-71f0e096e988.png)

### Design Improvements
There's a couple things I notice right away:
- There are too many various button states that are visible creating a hodge-podge of elements.
- The "undo/redo" items are actually requesting more attention because of their grey background than the simple blue text.
- "Block navigation" and "More options" take up too much space using two words. Making these changes can help:
    - Change "Outline" to "Details"
    - Change "Block navigation" to "Outline"
    - Change "More options" to "Options" (Keep it as "Options" regardless of whether "Settings" is included under it or not)
- "Add block" is too much text too... 
    - Change "Add block" to "Add"

### Desktop menu [prototype](https://www.figma.com/proto/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=234%3A270&viewport=-6704%2C-1092%2C1&scaling=min-zoom)
I haven't revisited the ordering of items in text-only mode yet to address the hierarchy and grouping issues, but this is moving closer to styling improvements on the existing state.

![iterated](https://user-images.githubusercontent.com/617986/91238670-6c335900-e6f2-11ea-8e74-acb366bc99db.gif)


### Dropdowns
The dropdowns need some iterations as well. I know you're thinking of saving this for another PR, but I thought documenting the changes here would be fine until that PR is created. :)

![dropdowns](https://user-images.githubusercontent.com/617986/91236642-58d1bf00-e6ed-11ea-8c6e-acf8737e3dce.png)

- The width of the popovers should be `260px` to remain consistent with the other popovers.
- When "Settings" drops into the "Options" dropdown, it loses it's black background state. I think this is fine as long as all buttons in the dropdown are consistent. Right now they aren't (ie. undo/redo).
- The "Options" dropdown no longer sits centered above the Sidebar settings panel. It's moved further to the left. It would be good to keep its positioning.
- All buttons in the dropdowns should have a consistent width of `234px`.
- When menu items get moved into dropdowns, their parent button has a blue border. This is not consistent with the other buttons like "Preview" when in the topbar. Can we remove the border?
- None of the popovers should overlap each other or be displayed below each other as mentioned above.

### Dropdown [prototype](https://www.figma.com/proto/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=476%3A1&viewport=-1545%2C625%2C0.5&scaling=min-zoom)
I iterated on some improvements for the dropdowns/popovers. I would imagine them working more closer to this:

![dropdowns](https://user-images.githubusercontent.com/617986/91241802-99840500-e6fa-11ea-908a-95b6789844bc.gif)

tellthemachines committed:
Change labels

tellthemachines committed:
Styling tweaks

tellthemachines committed:
Try Settings button inside Options dropdown

tellthemachines committed:
Expand modes in Tools dropdown

tellthemachines commented:
I think I've addressed all the feedback now, with the only pending items being the position of "Outline" and "Details" dropdowns. Those are dependent on upcoming Popover component API changes, so should be tackled separately.

@mapk I gave your dropdown prototype rearrangement a go, and it's working pretty well! Let me know what you think 😄

mapk commented:
Digging in a bit more, I've found a few design issues with the buttons.

### Button padding
The horizontal padding on both the "Add" and "Settings" buttons is too tight.

![Screen Shot 2020-08-26 at 4 28 24 PM](https://user-images.githubusercontent.com/617986/91366724-605e9a00-e7b9-11ea-9639-7d567836f5f4.png)

and

![Screen Shot 2020-08-26 at 4 28 32 PM](https://user-images.githubusercontent.com/617986/91366737-63f22100-e7b9-11ea-9f85-c05eaf7f1089.png)

### Button proximity
The "Options" button is sitting too close to the "Settings" button. It is visually odd. Can it have the same space between that the "Preview" and "Publish" buttons have?

![Screen Shot 2020-08-26 at 4 28 44 PM](https://user-images.githubusercontent.com/617986/91366814-9ef45480-e7b9-11ea-9c36-219bafecd2ab.png)

tellthemachines committed:
Button styling adjustments

tellthemachines committed:
Fix inserter aria label

tellthemachines committed:
Fix e2e selector.

tellthemachines committed:
Actually change all the e2e selectors.

tellthemachines commented:
@mapk all the button styling should be fixed now! 

I think this is ready for another code review too.

adamziel commented:
Thank you so much for taking on this @tellthemachines - I can see it's a grind, but it's going to help so many people!

A few visual glitches I found:

1. Notice how the buttons stack on iPhone 5. Nothing wrong with that per se, but they're also getting cut off:

<img width="383" alt="Zrzut ekranu 2020-08-27 o 11 02 56" src="https://user-images.githubusercontent.com/205419/91420699-52f6ed80-e855-11ea-86c9-b93a8346a460.png">
<img width="409" alt="Zrzut ekranu 2020-08-27 o 11 03 04" src="https://user-images.githubusercontent.com/205419/91420706-55594780-e855-11ea-875a-1ae0997b45fd.png">

2. The stacking causes some visual flickering when saving the post:

![2020-08-27 11-04-21 2020-08-27 11_04_35](https://user-images.githubusercontent.com/205419/91420876-846fb900-e855-11ea-8c7b-5b21ccf55e91.gif)

3. Probably unrelated to this specific PR, but it seems like icon buttons have a tad too large margin on mobile:

<img width="140" alt="Zrzut ekranu 2020-08-27 o 11 04 57" src="https://user-images.githubusercontent.com/205419/91420919-93ef0200-e855-11ea-8cca-e8249f7a5d1a.png">

4. Again that's probably unrelated to this specific PR, but I noticed that block toolbar has light background on mobile and transparent background on desktop:

<img width="1413" alt="Zrzut ekranu 2020-08-27 o 11 05 08" src="https://user-images.githubusercontent.com/205419/91420988-ab2def80-e855-11ea-87fe-7b6ad28ac023.png">

adamziel commented:
Also just pointing out - it seems super weird to me to have both "Settings" and "Options" buttons - I have no idea how they're different until I actually click both and then hopefully remember what they do after a few interactions with the editor:

<img width="173" alt="Zrzut ekranu 2020-08-27 o 11 09 41" src="https://user-images.githubusercontent.com/205419/91421199-ec260400-e855-11ea-8db1-4be6494f0803.png">

adamziel commented:
Let's compare Settings with Options top toolbar buttons:

<img width="720" alt="Zrzut ekranu 2020-08-27 o 11 42 08" src="https://user-images.githubusercontent.com/205419/91424806-91db7200-e85a-11ea-9078-a8921ac0d58e.png">

It didn't bother me as much when they were just icons, but now that they're text the inconsistencies are apparent. The same applies to tools dropdown.

Speaking of which, in the Options dropdown, all the menu items have black font as on the screenshot above. In the Tools dropdown, there's a mixture of blue and black fonts:

<img width="350" alt="Zrzut ekranu 2020-08-27 o 11 23 02" src="https://user-images.githubusercontent.com/205419/91422798-f0532100-e857-11ea-9b81-4e5b2ef85db2.png">

kevin940726 commented on packages/edit-post/src/components/header/header-toolbar/index.js:
```diff
@@ -89,22 +127,82 @@ function HeaderToolbar() {
+									choices={ [
+										{
+											value: 'edit',
+											label: <>{ __( 'Edit' ) }</>,
```
Nitpick/Question: Do we need to wrap this in a `Fragment`?

tellthemachines replied:
Whoops, no 😅

kevin940726 commented on packages/edit-post/src/components/header/style.scss:
```diff
@@ -116,3 +116,107 @@
+		// ... and display labels.
+		&::after {
+			content: attr(aria-label);
+			// margin: 0 0.5em;
```
Is this comment intentional?

tellthemachines replied:
Nope, removing!

kevin940726 commented:
![image](https://user-images.githubusercontent.com/7753001/91437801-06cf9b80-e89d-11ea-87c7-b452424a82b5.png)

Not sure if it's intentional to display the background color as transparent?

![image](https://user-images.githubusercontent.com/7753001/91437903-2a92e180-e89d-11ea-84ed-9199cd862597.png)

Also, if we have the "Autosave" banner at the same time, it will overlap with the toolbar.

![image](https://user-images.githubusercontent.com/7753001/91438261-a68d2980-e89d-11ea-96ec-48c39060785e.png)

In a smaller viewport, the toolbar is on top of the autosave banner.

ZebulanStanphill commented:
The ellipsis menu is a mess. We should probably look into cleaning it up in a separate PR. The stuff under "View" are really just editor preferences, so shouldn't they be inside the Options modal? And maybe that options modal should be renamed to "Preferences" or "Editor preferences".

The "Settings" button is interesting. It's vague name is ironically appropriate because it opens a sidebar that combines two things that should never have been combined in the first place: document settings and the block inspector. The only thing they both have in common is that they're both "settings". I think the only real solution here is to split the two apart. Really, the block inspector shouldn't even be accessed from the top toolbar in the first place; it should be accessed from the block toolbar, and it should act like a modal, to better fit accessibility guidelines.

If we can solve problems like these first, then I think it will be a lot easier to land this PR.

mapk commented:
>Having Options nested in Options does not help either:

Yeah, @adamziel, take a look at everywhere we use the word "Tools" too! 😆  https://github.com/WordPress/gutenberg/issues/19409

I believe a high level of reorganizing can help here. But that's an issue above and beyond this PR.

tellthemachines committed:
Styling improvements

tellthemachines committed:
Fix buttons on small breakpoint.

tellthemachines commented on packages/components/src/dropdown-menu/index.js:
```diff
@@ -129,7 +129,7 @@ function DropdownMenu( {
 						aria-expanded={ isOpen }
 						label={ label }
-						showTooltip
+						showTooltip={ toggleProps?.showTooltip ?? true }
```
Yes! Thanks for the suggestion 😄

tellthemachines committed:
Address remaining feedback and style fixes.

tellthemachines commented:
OK, toolbar transparency and mobile layout issues should be fixed now. I had to change the display text on the "Save as..." button to just "Save" on small breakpoint (mainly because "Save as pending" was breaking the layout no matter how much margin and padding were adjusted).

On "Settings" vs "Options" buttons styling, and on font color differences in the "Tools" dropdown, I will defer to @mapk as I've been following his designs.

Regarding the nested "Options" and the semantics of menu organisation more generally, all the points presented above by @adamziel and @ZebulanStanphill are valid, but solving them will require quite a bit of thinking and reorganising, so they shouldn't block this piece of work. Shipping a V1 of this will force future designs to take text labels into account, so hopefully some of the issues we're encountering here won't happen again 🙂

mtias commented:
I agree with @tellthemachines — the scope of this PR should be to land the option of having labeled buttons (with or without icon) and any trivial modifications to names that could be considered general improvements. Any further substantial changes to things like the inspector or the options + plugins menu is going to keep punting this improvement.

On the subject of text only and icon + text, I think it might be fine to include the three options (text, icons, text + icons).

Once we land this let's make sure to dedicate some time at reorganizing the settings modal, as it has grown a bit unconcernedly over the past couple months.

draganescu reviewed (approved):
I went through the code and it's 👍 (this PR went through a few code reviews so far) and I tested on browser and mobile. It works acceptable for this phase. The PR is big enough and it implements a change that can be a foundation to build upon.

P.S. I love text only buttons so I would live even the block toolbar to get these labels automatically when it's fixed to top :)

aristath reviewed (approved):
Code looks good to me and it makes sense. Tested and looks great :+1:

mapk commented:
Caught some spacing issues on smaller screens.

![spacing](https://user-images.githubusercontent.com/617986/91913492-9ba21280-ec6a-11ea-9553-4384b7f5fc2f.gif)

mapk commented:
Generally, the spacing in between these buttons all feels a bit off.

**These all feel too close.**

![Screen Shot 2020-09-01 at 3 50 10 PM](https://user-images.githubusercontent.com/617986/91913711-1f5bff00-ec6b-11ea-85ca-f7487c45efb6.png)

**These feel too far from one another.**

![Screen Shot 2020-09-01 at 3 50 19 PM](https://user-images.githubusercontent.com/617986/91913732-28e56700-ec6b-11ea-94a1-cd0d745dbcdf.png)

Is there a way to make this more consistent across all the buttons in text-only mode? In this Figma mockup, I have them spaced 10px apart.

![good-spacing](https://user-images.githubusercontent.com/617986/91914003-ef612b80-ec6b-11ea-9b4e-cf8f8e8ded94.png)

[Figma mockup](https://www.figma.com/file/ThRPgocCUUZmQJETnfPkT7/Icon-labels?node-id=690%3A52)

ZebulanStanphill commented:
`10px` isn't aligned to our [grid system](https://github.com/WordPress/gutenberg/blob/master/packages/base-styles/_variables.scss#L27), but yeah, something close to that would make sense.

ZebulanStanphill commented:
Noting that #24965 has been created to discuss improvements to the "Options" modal, which should help solve some of the overly-similar labels currently present.

See also these other issues relating to the top toolbar:
- #21870 G2: Apply 12px grid to top toolbar more consistently
- #20877 Rethinking the "Current Document" display and settings
- #20592 G2 block top toolbar
- #19409 Two distinct features now labeled "Tools"
- #19128 Exploration: moving options

talldan commented:
@mapk Was wondering why 'Outline' was used for the button text that opens the List View, instead of 'List View'?

Looking back through the designs in the comments, it looks like at one point it said 'List View' (https://github.com/WordPress/gutenberg/pull/24304#issuecomment-671609218) but then changed in the more recent designs.

mapk commented:
@ZebulanStanphill, changing the spacing to 12px is totally fine. 

@talldan, No problem changing it to "List View". Let's make that happen. :)

afercia commented:
Glad to see this merged! Thanks everyone for the perseverance and I'd like to thank @nicolad who started the first attempt to build this in #15830 ❤️ (deserves props).

Regarding this point by @ZebulanStanphill:

> Really, the block inspector shouldn't even be accessed from the top toolbar in the first place; it should be accessed from the block toolbar, and it should act like a modal, to better fit accessibility guidelines.

Ah! That's what the accessibility team has been asking in the last... years? Considering all the troubles for keyboard navigation to go from the selected block to the inspector and back, the teams asked several times to have the inspector "in place" close to the block instead of in the sidebar. Hopefully there will be new explorations in this direction.

afercia commented:
I'd like to propose props also for @tomjn who created [the original issue](https://github.com/WordPress/gutenberg/issues/10524), together with @nicolad who, as mentioned above, worked hard on the first PR.

tomjn commented:
I actually raised the issue during a user group after a discussion with @mikelittle, he deserves the props :)
